task/host-sp-comms/src/main.rs

```diff
@@ -301,6 +334,54 @@ impl ServerImpl {
         }
     }
 
+    fn handle_timer_notification(&mut self) {
+        let now = sys_get_timer().now;
+
+        // If we're past the deadline for transitioning to A0, attempt to do so.
+        if let Some(RebootState::TransitionToA0At(deadline)) = self.reboot_state
+        {
+            if now >= deadline {
+                // The only way our reboot state gets set to
+                // `TransitionToA0At` is if we believe we were currently in
+                // A2. Attempt to transition to A0, which can only fail if
+                // we're no longer in A2. In either case (we successfully
+                // started the transition or we're no longer in A2 due to
+                // some external cause), we've done what we can to reboot,
+                // so clear out `reboot_state`.
+                ringbuf_entry!(Trace::SetState {
+                    now: sys_get_timer().now,
+                    state: PowerState::A0,
+                });
+                _ = self.sequencer.set_state(PowerState::A0);
+                self.reboot_state = None;
+            }
+        }
+
+        // If we're past the deadline to send a 0x00 byte on the uart, do so and
+        // reset the timer, unless we're currently mid-send or mid-recv.
+        if let Some(deadline) = self.tx_zero_deadline {
+            if now >= deadline {
+                if self.tx_pkt_to_write.is_empty() && self.rx_buf.is_empty() {
+                    // If we have room in the FIFO, send a 0 byte and then reset
+                    // our timer to wake back up. If we can't, turn on the TX
+                    // FIFO empty interrupt instead, and wait for it to wake us
+                    // up.
+                    if try_tx_push(&self.uart, 0) {
+                        self.tx_zero_deadline = Some(now + UART_ZERO_DELAY);
+                        self.uart.disable_tx_fifo_empty_interrupt();
```

Comment:
I'm confused why we disable the FIFO empty interrupt here.  We've just pushed data into the FIFO, do we not want to do anything when it's sent?

Reply:
I think this is just a mistake that happens to not really matter when hardware flow control is working (because the tx fifo stays full until the host sends us a packet, at which point we take a uart interrupt due to rx instead of tx). I'll remove it - nice catch.